include/swift/AST/CaptureInfo.h

```diff
@@ -114,31 +116,63 @@ class DynamicSelfType;
 
 /// Stores information about captured variables.
 class CaptureInfo {
-  const CapturedValue *Captures;
-  DynamicSelfType *DynamicSelf;
-  OpaqueValueExpr *OpaqueValue;
-  unsigned Count = 0;
-  bool GenericParamCaptures : 1;
-  bool Computed : 1;
+  class CaptureInfoStorage final
+      : public llvm::TrailingObjects<CaptureInfoStorage, CapturedValue> {
+
+    DynamicSelfType *DynamicSelf;
+    OpaqueValueExpr *OpaqueValue;
+    unsigned Count;
+  public:
+    explicit CaptureInfoStorage(unsigned count, DynamicSelfType *dynamicSelf,
+                                OpaqueValueExpr *opaqueValue)
+      : DynamicSelf(dynamicSelf), OpaqueValue(opaqueValue), Count(count) { }
+
+    ArrayRef<CapturedValue> getCaptures() const {
+      return llvm::makeArrayRef(this->getTrailingObjects<CapturedValue>(),
+                                Count);
+    }
+
+    DynamicSelfType *getDynamicSelfType() const {
+      return DynamicSelf;
+    }
+
+    OpaqueValueExpr *getOpaqueValue() const {
+      return OpaqueValue;
+    }
+  };
+
+  enum class Flags : unsigned {
+    HasGenericParamCaptures = 1 << 0
+  };
+
+  llvm::PointerIntPair<const CaptureInfoStorage *, 2, OptionSet<Flags>>
+      StorageAndFlags;
 
 public:
-  CaptureInfo()
-    : Captures(nullptr), DynamicSelf(nullptr), OpaqueValue(nullptr), Count(0),
-      GenericParamCaptures(0), Computed(0) { }
+  /// The default-constructed CaptureInfo is "not yet computed".
+  CaptureInfo() = default;
+  CaptureInfo(ASTContext &ctx, ArrayRef<CapturedValue> captures,
+              DynamicSelfType *dynamicSelf, OpaqueValueExpr *opaqueValue,
+              bool genericParamCaptures);
 
-  bool hasBeenComputed() const { return Computed; }
+  /// A CaptureInfo representing no captures at all.
+  static CaptureInfo empty();
```

Comment:
In this PR `CaptureInfo::empty()` isn't so useful, but for the road of "capture info should always be explicit" it's a convenient shorthand.